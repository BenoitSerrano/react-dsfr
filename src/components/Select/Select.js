import { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';
// import '@gouvfr/selects/dist/css/selects.min.css';

/**
 * La liste déroulante permet à un utilisateur de choisir un élément dans une liste donnée.
 *
 * @visibleName Liste déroulante - Select
 */
const Select = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
}) => {
  const selectClasses = classNames(
    'rf-select',
    className,
    `${(messageType ? `rf-select--${messageType}` : '')}`,
  );
  const selectId = id || uuidv4();
  const messageId = uuidv4();
  const divClasses = classNames(`${(messageType ? `rf-select-group rf-select-group--${messageType}` : '')}`);

  return (
    <div className={divClasses}>
      {
      label && (
        <label className="rf-label" htmlFor={selectId} aria-describedby={messageId}>
          {label}
          {hint && <span className="rf-hint-text" id={`${selectId}-desc-hint`}>{hint}</span>}
        </label>
      )}
      <select
        className={selectClasses}
        data-testid="select"
        disabled={disabled}
        id={selectId}
        name={name}
        onChange={onChange}
      >
        {
          options.map((opt) => (
            <option
              disabled={opt.disabled || false}
              hidden={opt.hidden || false}
              key={uuidv4()}
              selected={opt.selected || false}
              value={opt.value}
            >
              {opt.label}
            </option>
          ))
        }
      </select>
      {(message && messageType) && <p id={messageId} className={`rf-${messageType}-text`}>{message}</p>}
    </div>
  );
};

Select.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: '',
  name: null,
  onChange: () => {},
};

Select.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    selected: PropTypes.bool,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
};

export default Select;
