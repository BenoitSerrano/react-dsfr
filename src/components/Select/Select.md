## Liste déroulante - Select

La liste déroulante permet à un utilisateur de choisir un élément dans une liste donnée.
### Liste déroulante
La liste déroule fournit une liste d’option parmi laquelle l’utilisateur peut choisir. Seule la partie visible du composant est stylisé : la liste d’option déroulée conserve le style du navigateur. 