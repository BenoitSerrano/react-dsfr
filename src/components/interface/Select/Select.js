import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import SelectWrapper from './SelectWrapper';

/**
 *
 * @visibleName Select
 */
const Select = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
  selected,
  required,
  ...remainingProps
}) => {
  const _className = classNames('fr-select', {
    [`fr-select--${messageType}`]: messageType,
  });
  const selectId = useRef(id || uuidv4());

  useEffect(() => {
    selectId.current = id || uuidv4();
  }, [id]);

  return (
    <SelectWrapper
      className={className}
      hint={hint}
      selectId={selectId.current}
      label={label}
      message={message}
      messageType={messageType}
      required={required}
      {...remainingProps}
    >
      <select
        className={_className}
        disabled={disabled}
        id={selectId.current}
        name={name}
        onChange={onChange}
        value={selected}
        data-testid="select"
      >
        {
          options.map((opt) => (
            <option
              disabled={opt.disabled || false}
              hidden={opt.hidden || false}
              key={`${selectId.current}-${opt.value}`}
              value={opt.value}
            >
              {opt.label}
            </option>
          ))
        }
      </select>
    </SelectWrapper>
  );
};

Select.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: undefined,
  name: null,
  onChange: () => {},
  selected: '',
  required: false,
};

Select.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  required: PropTypes.bool,
};

export default Select;
