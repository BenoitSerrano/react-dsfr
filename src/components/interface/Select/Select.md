### Liste déroulante

```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
  />
```

### État d'erreur
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
    message="Texte d'erreur obligatoire"
    messageType="error"
  />
```
### État de succès
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
    message="Texte de validation"
    messageType="valid"
  />
```
### État inactif
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    options={options}
    disabled={true}
  />
```
### Liste déroulante avec texte d’aide
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true},
    {value: "1", label: "Burgers & sandwiches"}, {value: "2", label: "Build your own"},
    {value: "3", label: "Friendly meals"},
    {value: "4", label: "Save menu & more"}
  ];
  <Select
    label="Label pour liste déroulante"
    hint="Texte de description additionnel"
    options={options}
  />
```
