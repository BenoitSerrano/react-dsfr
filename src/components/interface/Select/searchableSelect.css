.select-search-options {
    background-color: var(--w);
    box-shadow: 0px 0px 10px var(--g300);
    position: absolute;
    z-index: 10;
    width: 100%;
    max-height: 32vh;
    overflow: scroll;
    display: none;
}

.select-search-options__visible {
    display: inherit;
}

.select-search-option__disabled,
.select-search-option {
    padding: 8px 16px;
    color: var(--bf500);
}

.select-search-option {
    cursor: pointer;
}

.select-search-option:hover {
    background-color: var(--block-color-hover);
}
