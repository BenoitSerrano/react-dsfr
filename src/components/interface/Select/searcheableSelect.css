.select-search-options {
    background-color: white;
    box-shadow: 0px 4px 30px rgba(43, 43, 43, 0.13);
    position: absolute;
    z-index: 10;
    width: 100%;
    max-height: 32vh;
    overflow: scroll;
    display: none;
}

.select-search-options__visible {
    display: inherit;
}

.select-search-option__disabled,
.select-search-option {
    padding: 8px 16px;
    color: var(--bf500);
}

.select-search-option {
    cursor: pointer;
}

.select-search-option:hover {
    background-color: var(--block-color-hover);
}
