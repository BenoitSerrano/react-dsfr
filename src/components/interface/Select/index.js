import SearcheableSelect from './SearcheableSelect';
import Select from './Select';

export {
  SearcheableSelect,
  Select,
};
