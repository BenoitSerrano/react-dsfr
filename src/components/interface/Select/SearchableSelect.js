import React, {
  createRef, useEffect, useRef, useState,
} from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import SelectWrapper from './SelectWrapper';
import './searchableSelect.css';

/**
 *
 * @visibleName SearchableSelect
 */
const SearchableSelect = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
  selected,
  required,
  filter,
  ...remainingProps
}) => {
  const selectId = useRef(id || uuidv4());
  const optionsRef = useRef([]);
  const optionsContainerRef = useRef();
  const [arrowSelected, setArrowSelected] = useState();

  const getSelectedLabel = () => {
    if (selected) {
      const selectedOption = options.find((option) => option.value === selected);
      return selectedOption ? selectedOption.label : '';
    }
    return '';
  };

  const [internalLabel, setInternalLabel] = useState(getSelectedLabel());
  const [showOptions, setShowOptions] = useState(false);

  if (options.length !== optionsRef.length) {
    optionsRef.current = Array(options.length)
      .fill()
      .map((option, i) => optionsRef.current[i] || createRef());
  }

  useEffect(() => {
    selectId.current = id || uuidv4();
  }, [id]);

  useEffect(() => {
    if (arrowSelected) {
      optionsContainerRef.current.scrollTop = Math.max(
        0,
        optionsRef.current[arrowSelected].current.offsetTop - 20,
      );
    } else {
      optionsContainerRef.current.scrollTop = 0;
    }
  }, [arrowSelected]);

  const _className = classNames('fr-select', {
    [`fr-select--${messageType}`]: messageType,
  });

  const filteredOptions = options
    .filter((option, index, arr) => filter(internalLabel, option, index, arr));

  const onInternalChange = (newValue, newLabel) => {
    setInternalLabel(newLabel);
    onChange(newValue);
  };

  const onFocus = () => {
    onInternalChange('', '');
    setShowOptions(true);
    setArrowSelected(null);
  };

  const onBlur = () => {
    if (filteredOptions.length === 1) {
      onInternalChange(filteredOptions[0].value, filteredOptions[0].label);
    } else {
      const foundValue = options.find((option) => option.label === internalLabel);
      if (!foundValue) {
        setInternalLabel('');
      }
    }
    setShowOptions(false);
  };

  const onKeyDown = (e) => {
    switch (e.key) {
      case 'ArrowDown':
        e.preventDefault();
        setShowOptions(true);
        if (arrowSelected === null) {
          setArrowSelected(0);
        } else if (arrowSelected < filteredOptions.filter((o) => !o.disabled).length - 1) {
          setArrowSelected(arrowSelected + 1);
        }
        break;
      case 'ArrowUp':
        e.preventDefault();
        setShowOptions(true);
        if (arrowSelected && arrowSelected > 0) {
          setArrowSelected(arrowSelected - 1);
        }
        break;
      case 'Enter':
        e.preventDefault();
        if (arrowSelected !== null) {
          const option = filteredOptions.filter((o) => !o.disabled)[arrowSelected];
          onInternalChange(option.value, option.label);
          setShowOptions(false);
        }
        break;
      default:
        setArrowSelected(null);
    }
  };

  let refCount = -1;
  return (
    <SelectWrapper
      className={className}
      hint={hint}
      selectId={selectId.current}
      label={label}
      message={message}
      messageType={messageType}
      required={required}
      {...remainingProps}
    >
      <div className="select-search-input">
        <input
          onKeyDown={onKeyDown}
          id={selectId.current}
          className={_className}
          autoComplete="off"
          required={required}
          disabled={disabled}
          onChange={(e) => setInternalLabel(e.target.value)}
          onFocus={onFocus}
          onBlur={onBlur}
          value={internalLabel}
        />
        <div
          ref={optionsContainerRef}
          className={classNames(
            'select-search-options',
            'midlength-input',
            { 'select-search-options__visible': showOptions },
          )}
        >
          {filteredOptions.length === 0 ? (
            <div className="select-search-option__disabled">
              Aucun r√©sultat
            </div>
          ) : (
            <>
              {filteredOptions.map((option) => {
                if (!option.disabled) {
                  refCount += 1;
                }
                return (
                  <option
                    ref={option.disabled ? null : optionsRef.current[refCount]}
                    className={classNames(
                      'select-search-option',
                      {
                        'select-search-option__selected': !option.disabled && refCount === arrowSelected,
                        'select-search-option__disabled': option.disabled,
                      },
                    )}
                    disabled={option.disabled || false}
                    hidden={option.hidden || false}
                    key={`${selectId}-${option.value}`}
                    value={option.value}
                    onMouseDown={() => onInternalChange(option.value, option.label)}
                  >
                    {option.label}
                  </option>
                );
              })}
            </>
          )}
        </div>
      </div>
    </SelectWrapper>
  );
};

SearchableSelect.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: undefined,
  name: null,
  onChange: () => {},
  selected: '',
  required: false,
  filter: (
    label,
    option,
  ) => option.label.toLowerCase().includes(label.toLowerCase()),
};

SearchableSelect.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.string,
  required: PropTypes.bool,
  filter: PropTypes.func,
};

export default SearchableSelect;
