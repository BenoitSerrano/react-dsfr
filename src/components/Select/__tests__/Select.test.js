import { render, screen } from '@testing-library/react';
import Select from '..';

describe('<Select />', () => {
  const options = [
    { value: '1', label: 'label-1' },
    { value: '2', label: 'label-2' },
  ];

  it('renders Select properly', () => {
    // const onChange = jest.fn();
    render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    const search = screen.getByTestId('select');
    expect(search).toBeInTheDocument();
    // expect(search).toMatchSnapshot();
  });

  // test du nombre d'items

  // test de la présence de la class rf-select

  // test du label

  // test de la fonction onChange

  // test de la class d'erreur
  // label
  // select
  // message
  // aria

  // test de la class succès
  // label
  // select
  // message
  // aria

  // test disable

  // test hint
  // ...
});
