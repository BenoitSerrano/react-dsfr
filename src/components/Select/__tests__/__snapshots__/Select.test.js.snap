// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders Select properly 1`] = `
<select
  class="rf-select"
  data-testid="select"
  id="591585d9-723a-4aac-a097-439192fb9aa8"
>
  <option
    value="1"
  >
    label-1
  </option>
  <option
    value="2"
  >
    label-2
  </option>
</select>
`;
